lib/api.c

```diff
@@ -78,6 +83,186 @@ EXP int get_bios_if_ver(ryzen_access ry)
+		printf("Get table version is unsupported\n");
+		return 0;
+	} else {
+		printf("Get table version was rejected\n");
+		return 0;
+	}
+}
+
+EXP size_t get_table_size(ryzen_access ry)
+{
+	if(ry->table_size)
+		return ry->table_size;
+	
+	switch (get_table_ver(ry))
+	{
+		case 0x1E0004: ry->table_size = 0x6AC; break;
+		case 0x1E0005: ry->table_size = 0x6AC; break;
+		case 0x1E0101: ry->table_size = 0x6AC; break;
+		case 0x240802: ry->table_size = 0x7E0; break;
+		case 0x240803: ry->table_size = 0x7E4; break;
+		case 0x240902: ry->table_size = 0x514; break;
+		case 0x240903: ry->table_size = 0x518; break;
+		case 0x2D0803: ry->table_size = 0x894; break;
+		case 0x380804: ry->table_size = 0x8A4; break;
+		case 0x2D0903: ry->table_size = 0x594; break;
+		case 0x380904: ry->table_size = 0x5A4; break;
+		case 0x370000: ry->table_size = 0x794; break;
+		case 0x370001: ry->table_size = 0x884; break;
+		case 0x370002: ry->table_size = 0x88C; break;
+		case 0x370003: ry->table_size = 0x8AC; break;
+		case 0x370004: ry->table_size = 0x8AC; break;
+		case 0x370005: ry->table_size = 0x8C8; break;
+		default:
+			//use smallest size for unknown version to support dump of unknown tables
+			ry->table_size = 0x514;
+	}
+	return ry->table_size;
+}
+
+EXP uint32_t get_table_addr(ryzen_access ry)
+{
+	if(ry->table_addr)
+		return ry->table_addr;
+
+	unsigned int get_table_addr_msg;
+	smu_service_args_t args = {0, 0, 0, 0, 0, 0};
+	switch (ry->family)
+	{
+	case FAM_RAVEN:
+	case FAM_PICASSO:
+		args.arg0 = 3;
+		get_table_addr_msg = 0xB;
+		break;
+	case FAM_RENOIR:
+	case FAM_CEZANNE:
+		get_table_addr_msg = 0x66;
+		break;
+	default:
+		printf("Get table addr is not supported on this family\n");
+		return 0;
+	}
+
+	int resp = smu_service_req(ry->psmu, get_table_addr_msg, &args);
+	if (resp == REP_MSG_OK) {
+		ry->table_addr = args.arg0;
+		if(!ry->table_addr)
+		{
+			printf("Get table addr did not return anything\n");
+		}
+	} else if (resp == REP_MSG_UnknownCmd) {
+		printf("Get table addr is unsupported\n");
+		return 0;
+	} else {
+		printf("Get table addr was rejected\n");
+		return 0;
+	}
+
+	//init memory object because it is prerequiremt to woring with physical memory address
+	ry->mem_obj = init_mem_obj();
+	if(!ry->mem_obj)
+	{
+		printf("Unable to get MEM Obj\n");
+		return 0;
+	}
+	
+	//copy table after finding the address to have a fresh table for get value calls
+	//avoids checking for existing table in each get value call
+	if(!refresh_pm_table(ry))
+	{
+		printf("Unable to get power table\n");
+		return 0;
+	}
+	
+	uint32_t data;
+	physical_memory_read_u32(ry->table_addr, &data);
+	if(!data){
+		//Raven don't get table refresh on the very first transfer call after boot
+		//if we detact 0 data, do an initial 2nd call after a small delay (physical_memory_read is enough delay)
+		//transfer, transfer, wait, wait longer; don't work
+		//transfer, wait, wait longer; don't work
+		//transfer, wait, transfer; does work
+		refresh_pm_table(ry);
+	}
```

Comment:
@FlyGoat did you see this?
At least I could avoid the unused getTableAdress commands 
A + arg0: 3
A + arg0: 5
B + arg0: 5

But we still need this short solution of the "Wait for table" thing for transfer table. I hope this will be enough, otherwise we need to implement a loop with timeout.
Did work so far on 2400G, 2500U and 3700U